# Homework Introduction to Python

$\textbf{Ubah nama di bawah yaa!}$

- Nama: Tegas Gagas Impian Prahita N, S.Si, M.Si.
- IPK / Nilai Rapot: 3.95 / 9.8

In [1]:
import numpy as np

# please import as much as you need

# Normal

## 1. Operasi Variabel

> Misal kita memiliki toko sepatu. Modal membeli 1 kodi sepatu adalah Rp 1.000.000. Setiap sepatu dijual seharga Rp 80.000. Buatlah variabel modal, kuantitas, dan harga_jual sesuai informasi di atas. Hitung profit yang didapatkan jika semua sepatu berhasil terjual, dan masukkan ke variabel profit.

Saran langkah-langkah pengerjaan
1. Convert terlebih dahulu kodi dalam jumlah barang (1 kodi = xyz buah sepatu)
2. Gunakan operasi matematika sederhana untuk menghitung profit menggunakan tiga buah variabel yang sudah ditentukan

In [2]:
# your code goes here!

#1
modal = 1e6
kuantitas = 20
harga_jual = 8e4

profit = harga_jual * kuantitas - modal
print(f"profit yang didapat yaitu Rp.{profit}")

profit yang didapat yaitu Rp.600000.0


Jadi, dari profit di atas, dapat disimpulkan bahwa Toko Sepatu mengalami keuntungan Rp.600.000,-

## 2. Package Python dan Dictionary

> Kita akan coba mengakses data scientific article dari PLOS corpus
    Menggunakan API request dengan kode berikut : 
```
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
```
**perhatikan tanda kutipnya tidak salah ketika meng-copy ke dalam python

Pertanyaannya : 
Convert response.content dari request tersebut dari JSON Format ke dalam Dictionary, dan print-kan key di dalamnya apa saja

Saran langkah-langkah pengerjaan
1. Ikuti code dari soal, kemudian ekstrak value dari variable response menggunakan dot (.) content menjadi sebuah variabel baru (dictionary)
2. Gunakan library json, convert hasil content menggunakan json.loads
3. Print keys dari dictionary tersebut


In [3]:
# your code goes here!
import json
import requests

r = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
print(r)

dictionary = json.loads(r.content)

dictionary.keys()

print(f"Isi keys dari 'Dictionary' :{list(dictionary.keys())}")

<Response [200]>
Isi keys dari 'Dictionary' :['response']


Jadi, dari url link di atas, kami menyimpulkan bahwa memiliki keys/columns response

# Intermediate

## 3. Operasi dalam List

> Buatlah list dengan 10 nama kota di indonesia, kemudian:
1. Urutkan nama kota di dalam list berdasarkan abjad (a)
2. Ambil 4 kota pertama dari list tersebut (yang telah diurutkan) (b)
3. Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan) (c)

Saran langkah-langkah pengerjaan
1. Tidak ada ketentuan nama kota (misal berdasarkan provinsi X atau pulau Y), dibebaskan untuk memilih. Buat nama-nama kota tersebut dalam sebuah list.
2. Gunakan fungsi sorted(X) untuk mengurutkan secara kecil-ke-besar (sesuai abjad), dimana X adalah contoh variabel list pada langkah (a)
3. Gunakan hasil poin (a), urutkan nama kota di dalam list berdasarkan abjad, untuk mengerjakan poin (b) dan (c). 

In [4]:
# your code goes here!

nama_kota = ['Jakarta','Bandung','Depok','Fakfak','Ende','Halmahera tengah','Indramayu','Ambon','Cirebon','Gorontalo']

# (a) Urutkan nama kota di dalam list berdasarkan abjad
a = sorted(nama_kota)
print(f"Hasil urutan nama kota berdasarkan abjad {a}")

# (b) Ambil 4 kota pertama dari list tersebut (yang telah diurutkan)
b = a[:4]
print(f"Hasil 4 kota pertama {b}")


# (c) Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan)
c = a[-3:]
print(f"Hasil 3 kota terakhir {c}")

Hasil urutan nama kota berdasarkan abjad ['Ambon', 'Bandung', 'Cirebon', 'Depok', 'Ende', 'Fakfak', 'Gorontalo', 'Halmahera tengah', 'Indramayu', 'Jakarta']
Hasil 4 kota pertama ['Ambon', 'Bandung', 'Cirebon', 'Depok']
Hasil 3 kota terakhir ['Halmahera tengah', 'Indramayu', 'Jakarta']


Jadi, dari operasi list di atas, dapat disimpulkan bahwa urutan nama kota yaitu Ambon, Bandung, Cirebon, Depok, Ende, Fakfak, Gorontalo, Halmahera tengah, Indramayu, Jakarta.

## 4. Function dalam List

> Team sales memiliki list total transaksi dari 10 merchant pada bulan Juni 2020, yaitu: 
```
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
```
1. Trx minimum
2. Rata-rata 
3. Trx maksimal

Saran langkah-langkah pengerjaan
1. Copy variabel total_transaksi yang ada di soal ke notebook kita (atau ketik ulang boleh)
2. Kemudian, hitung menggunakan function operasi dalam list yang telah dipelajari di kelas pada topik list. Print hasilnya setiap poin (a), (b), dan (c)

In [5]:
# your code goes here!

total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]

# Trx minimum
transaksi_paling_rendah = min(total_transaksi)
print(f"Transaksi paling rendah yaitu {transaksi_paling_rendah}")

# Rata-rata
rerata_transaksi = sum(total_transaksi)/len(total_transaksi)
print(f"Rata-rata transaksi yaitu {rerata_transaksi}")

# Trx maksimal
transaksi_paling_tinggi = max(total_transaksi)
print(f"Transaksi paling tinggi yaitu {transaksi_paling_tinggi}")


Transaksi paling rendah yaitu 10
Rata-rata transaksi yaitu 25.6
Transaksi paling tinggi yaitu 45


Jadi, dari hasil operasi list di atas, dapat disimpulkan bahwa nilai minimal, maksimal, dan rata-rata yaitu 10, 45, dan 25.6

# Advanced

## 5. NumPy Array

> Buatlah array dari 5 bilangan ganjil pertama (dimulai dari 1), Kemudian lakukan perintah berikut : 
1. Kali-kan array tersebut dengan 10
2. Bagi-kan array (hasil dari bagian a) dengan 3
3. Kali-kan array (hasil dari bagian b) dengan array baru :
`[-45, 43, 3, 56, 5]`
4. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c) 

Saran langkah-langkah pengerjaan
1. HATI-HATI dalam pengerjaan, karena apabila poin sebelumnya salah, maka selanjutnya akan kebawa salah (misal (b) salah, maka poin (c) sampai (d) akan salah)
2. Untuk array 5 bilangan ganjil pertama boleh manual boleh menggunakan fungsi, dimulai dari 1
3. Ikuti perintah soal, gunakan numpy array, jangan list. Sesuaikan dengan permintaan soal pada setiap butir.



In [6]:
# your code goes here!

odd_num = np.arange(1,10,2)

# Kali-kan array tersebut dengan 10
a = odd_num * 10
print(a)

# Bagi-kan array (hasil dari bagian a) dengan 3
b = a / 3
print(b)

# Kali-kan array (hasil dari bagian b) dengan array baru : [-45, 43, 3, 56, 5]
new = np.array([-45, 43, 3, 56, 5])
c = b * new
print(c)

# Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c)
print(f"Indeks nilai tertinggi yaitu {np.argmax(c)} dengan nilai {c[np.argmax(c)]}")
print(f"Indeks nilai terendah yaitu {np.argmin(c)} dengan nilai {c[np.argmin(c)]}")

[10 30 50 70 90]
[ 3.33333333 10.         16.66666667 23.33333333 30.        ]
[-150.          430.           50.         1306.66666667  150.        ]
Indeks nilai tertinggi yaitu 3 dengan nilai 1306.6666666666665
Indeks nilai terendah yaitu 0 dengan nilai -150.0


Jadi, dari hasil operasi numpy di atas, dapat disimpulkan bahwa nilai terendah dan tertinggi yaitu -150.0 dan 1306.667 pada indeks 0 dan 3

# Terimakasih, salam dari bandung